### PR TITLE
+ add mb_xdp prog for ingress packet

### PR DESCRIPTION
**Pull request by kebe7jun**
https://github.com/merbridge/merbridge/issues/116 for xdp prog

### CODE REVIEW

File: bpf/mb_xdp.c [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        return XDP_PASS;
+    }
+
+    struct tcphdr *tcph = (struct tcphdr *)(iph + 1);
+    if ((void *)(tcph + 1) > data_end) {
+        return XDP_DROP;
+    }
+    if (tcph->syn && !tcph->ack) {
+        // first packet
+        if (bpf_htons(tcph->dest) == IN_REDIRECT_PORT) {
```
Comment (dddddai) [inline bpf/mb_xdp.c:85]: Just a nit, would it be better to use `tcph->dest == bpf_htons(IN_REDIRECT_PORT)`? I'm thinking if the compiler would do some optimization at compile time

Comment (dddddai) [inline bpf/mb_xdp.c:85]: Just a nit, would it be better to use `tcph->dest == bpf_htons(IN_REDIRECT_PORT)`? I'm thinking if the compiler would do some optimization at compile time

File: bpf/mb_xdp.c [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+            .sport = tcph->source,
+            .dip = iph->daddr,
+            .dport = bpf_htons(IN_REDIRECT_PORT),
+        };
+        struct origin_info origin = {
+            .ip = iph->daddr,
+            .port = tcph->dest,
+            .flags = XDP_ORIGIN_FLAG,
+        };
+        bpf_map_update_elem(&pair_original_dst, &p, &origin, BPF_NOEXIST);
```
Comment (dddddai) [inline bpf/mb_xdp.c:136]: Should use `BPF_ANY`?

File: bpf/mb_xdp.c [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        // todo delete key
+        struct pair p = {
+            .dip = iph->saddr,
+            .dport = tcph->source,
+            .sip = iph->daddr,
+            .sport = bpf_htons(IN_REDIRECT_PORT),
+        };
+        bpf_map_delete_elem(&pair_original_dst, &p);
+        return XDP_PASS;
+    } else if (bpf_htons(tcph->source) == IN_REDIRECT_PORT) {
```
Comment (dddddai) [inline bpf/mb_xdp.c:151]: the same as above

Comment (dddddai) [inline bpf/mb_xdp.c:151]: the same as above

**Review by dddddai — COMMENTED** [review]
Cool!

File: bpf/mb_xdp.c [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    } else if (tcph->fin && tcph->ack) {
+        // todo delete key
+        struct pair p = {
+            .dip = iph->saddr,
+            .dport = tcph->source,
+            .sip = iph->daddr,
+            .sport = bpf_htons(IN_REDIRECT_PORT),
+        };
+        bpf_map_delete_elem(&pair_original_dst, &p);
+        return XDP_PASS;
```
Comment (dddddai) [inline bpf/mb_xdp.c:150]: Should we rewrite dst/src port when client/server closing the connection?

Comment (kebe7jun) [reply bpf/mb_xdp.c:150]: request and response logic have covered this, we do not need do this manually.

**Comment by kebe7jun** [comment]
@dddddai updated.

**Review by dddddai — APPROVED** [review]

**Pull request merged**